server/controllers/chatController.js

```diff
@@ -0,0 +1,221 @@
+                }),
+              },
+            },
+          },
+          {
+            $group: {
+              _id: {
+                messages: "$messages",
+                messagesId: "$_id",
+              },
+            },
+          },
+        ]);
+        const lastMessagesList = messageIds.map(item => {
+          return {
+            chatId: item.chatId.toString(),
+            companionUserId: item.companionUserId,
+            messageId: item.messageId.toString(),
+            lastMessage: lastMessage
+              .filter(element => {
+                return (
+                  element._id.messagesId.toString() ===
+                  item.messageId.toString()
+                );
+              })[0]
+              ._id.messages.slice(-1)[0],
+          };
+        });
+
+        //get list of all conversations
+        let companions = chats
+          .map(chat => {
+            return chat.participants.filter(participant => {
+              return participant !== req.user.id;
+            });
+          })
+          .flat();
+        //find Profile Id for every companion
+        const userIds = companions.map(id => {
+          return mongoose.Types.ObjectId(id);
+        });
+        let userProfiles = await User.aggregate([
+          { $match: { _id: { $in: userIds } } },
+          { $group: { _id: { profileId: "$profile", userId: "$_id" } } },
+        ]);
+        //find Profile Data for every companion
+        const profileIds = userProfiles.map(item => {
+          return mongoose.Types.ObjectId(item._id.profileId);
+        });
+        let profilesData = await Profile.aggregate([
+          { $match: { _id: { $in: profileIds } } },
+          {
+            $group: {
+              _id: {
+                firstName: "$firstName",
+                lastName: "$lastName",
+                profile: "$_id",
+                picture: "$profilePicture",
+              },
+            },
+          },
+        ]);
+        //Bring all the data together and create array of objects
+        let conversations = profilesData.map(item => {
+          return {
+            firstName: item._id.firstName,
+            lastName: item._id.lastName,
+            picture: item._id.picture,
+            profileId: item._id.profile,
+            userId: userProfiles.filter(element => {
+              return (
+                element._id.profileId.toString() == item._id.profile.toString()
+              );
+            })[0]._id.userId,
+          };
+        });
+        conversations.map(item => {
+          item.lastMessage = lastMessagesList.filter(element => {
+            return (
+              element.companionUserId.toString() === item.userId.toString()
+            );
+          })[0].lastMessage;
+          item.chatId = lastMessagesList.filter(element => {
+            return (
+              element.companionUserId.toString() === item.userId.toString()
+            );
+          })[0].chatId;
+        });
```

Comment:
There are a few implicit nested for loops (Array.filter inside of Array.map) in this approach, this would not scale well. We can either refactor now or leave as is for the time being (since it's working) and make a refactoring task to be done later

Reply:
I'd leave for now and refactor if we got time later